Develop

Nesailormun opened the pull request:
Выполнен групповой проект, реализованы функциональности, обозначенные в ТЗ.

b1mgd committed:
Add remove endpoint for films and users (#14)

Nesailormun committed:
Добавлена функциональность «Популярные фильмы» (#13)

* Обновлен метод getTopFilms в цепочке FilmController->FilmService->FilmDbStorage->FilmRepository согласно новой функциональности (фильтрации по genreId и year)

* Обновлен метод getTopFilms в цепочке FilmController->FilmService->FilmDbStorage->FilmRepository согласно новой функциональности (фильтрации по genreId и year)

* делаю коммит псÐоскольку force push не триггерит пайплайн по всей видимости

* Исправил логику работы метода getTopFilms(...) (параметры строки запроса genreId and year являются необязательными)

---------

Co-authored-by: nesailormun <Keyko07@yandex.ru>

1Boolean1 committed:
add getCommonFilms for user and friend (#10)

* add getCommonFilms for user and friend

* little change

* checkstyle

1Boolean1 committed:
Add remove endpoint (#17)

* add recommendations

* Реализован функционал добавления-исправления режиссёра, получения списка режиссёров по id, получение списка всех режиссёров. Реализовано добавление/исправление/получение фильмов с учетом информации о режиссере.

* Добавлен функционал возвращения списка фильмов режиссера отсортированных по лайкам или году выпуска

* Исправить ошибки Checkstyle

* Отладка тестов Postman: переименовать поле director в directors; внести правки в функционал исправления фильма в части информации о режиссере

* Переименовать метод deleteFilms в deleteFilmDirectors в классе FilmRepository. Добавить проверку на корректность введённых параметров в контроллере FilmController в методе findByDirector

* Add remove endpoint for films and users (#14)

* Добавлена функциональность «Популярные фильмы» (#13)

* Обновлен метод getTopFilms в цепочке FilmController->FilmService->FilmDbStorage->FilmRepository согласно новой функциональности (фильтрации по genreId и year)

* Обновлен метод getTopFilms в цепочке FilmController->FilmService->FilmDbStorage->FilmRepository согласно новой функциональности (фильтрации по genreId и year)

* делаю коммит псÐоскольку force push не триггерит пайплайн по всей видимости

* Исправил логику работы метода getTopFilms(...) (параметры строки запроса genreId and year являются необязательными)

---------

Co-authored-by: nesailormun <Keyko07@yandex.ru>

* add getCommonFilms for user and friend (#10)

* add getCommonFilms for user and friend

* little change

* checkstyle

* add getCommonFilms for user and friend

* Внести исправление в метод getTopFilms класса FilmRepository: join с таблицей likes заменить на left join

---------

Co-authored-by: Plastinin-Igor <igor-plastinin@yandex.ru>
Co-authored-by: b1mgd <sparcman1990@gmail.com>
Co-authored-by: Stanislav Mun <146199590+Nesailormun@users.noreply.github.com>
Co-authored-by: nesailormun <Keyko07@yandex.ru>

Plastinin-Igor committed:
Реализовать Функциональность Поиск

Plastinin-Igor committed:
Исправить ошибку Checkstyle

Commit:
Добавлена функциональность получения ленты событий пользователя при работе с друзьями и лайками. Ожидаю реализации функциональности review.

Plastinin-Igor committed:
Merge pull request #18 from 1Boolean1/add-search

Функциональность Поиск

b1mgd committed:
Add review endpoints

Commit:
Merge branch 'add-reviews' into develop

Commit:
Merge branch 'develop' into add-feed

# Conflicts:
#	src/main/resources/schema.sql

Nesailormun committed:
Merge pull request #19 from 1Boolean1/add-reviews

Add review endpoints

Commit:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

Commit:
Merge branch 'add-feed' into develop

Commit:
Доработана функциональность "Лента новостей" после добавления функциональности "Отзывы"

Nesailormun committed:
Merge pull request #21 from 1Boolean1/add-feed

Add feed

Plastinin-Igor committed:
Исправить ошибки Postman

Plastinin-Igor committed:
Исправить ошибку в тесте FilmorateApplicationTests

Commit:
Исправлены ошибки тестов Postman, связанные с add-review (update Review)

Nesailormun committed:
Merge pull request #23 from 1Boolean1/add-feed

Исправить ошибки Postman

avfyodorov commented on src/main/resources/application.properties:
```diff
@@ -1 +1,11 @@
+logging.level.org.zalando.logbook=INFO
+logging.level.ru.yandex.practicum.filmorate=INFO
+spring.sql.init.mode=always
+
```
Просьба добавить описание того, какие задачи выполнила группа, в описании пул-реквеста или в файле readme

1Boolean1 replied:
Добавили

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dto/UpdateFilmRequest.java:
```diff
@@ -1,6 +1,7 @@
 import lombok.Data;
+import ru.yandex.practicum.filmorate.model.Director;
 import ru.yandex.practicum.filmorate.model.Rating;
 
```
Насколько я понял, этот класс в проекте не используется, его можно спокойно убрать.

1Boolean1 replied:
Спасибо, исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/exception/InternalServerException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.yandex.practicum.filmorate.exception;
+
```
`public class ConstraintViolationException extends RuntimeException {`

Пишу здесь, потому что класс недоступен для комментария. 
Сделано вот такое самодельное исключение. Вопрос, есть же точно такое же , почему нельзя было воспользоваться стандартным? 

import jakarta.validation.ConstraintViolationException;

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/exception/InternalServerException.java:
```diff
@@ -0,0 +1,7 @@
+    public InternalServerException(String message) {
+        super(message);
+    }
+}
```
`public class MethodArgumentNotValidException extends RuntimeException {`

аналогичный вопрос????

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/FeedRepository.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+@Repository
+public class FeedRepository {
```
Вопрос-почему этот репозитарий нельзя было наследовать от базового? 

`public class BaseRepository<T> {`

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/ReviewRepository.java:
```diff
@@ -0,0 +1,142 @@
+Не стал оформлять ReviewDbStorage, так как в остальной программе такие классы лишь делигируют вызовы классу-репозиторию.
+В моем понимании ReviewDbStorage по заданию предыдущего спринта и есть ReviewRepository и т.п.
+ */
+
```
```
public class GenreRepository extends BaseRepository<Genre> {
............
    private final JdbcTemplate jdbc;
    private final GenreRowMapper genreMapper;
```

Класс недоступен для комментария, поэтому пишу здесь. 
А зачем здесь эти поля? Он же наследуется от базового репозитария... там уже есть.

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -30,12 +36,66 @@ public class FilmRepository extends BaseRepository<Film> {
+             WHERE
+            """;
 
     private final JdbcTemplate jdbc;
```
private final JdbcTemplate jdbc;
    private FilmRowMapper filmMapper;

Тот же вопрос - действительно ли нужны эти поля, если класс наследуется от базового репозитария?

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -30,12 +36,66 @@ public class FilmRepository extends BaseRepository<Film> {
+            """;
 
     private final JdbcTemplate jdbc;
     private final NamedParameterJdbcTemplate namedJdbcTemplate;
```
private final JdbcTemplate jdbc;  
private NamedParameterJdbcTemplate namedJdbcTemplate;

Действительно ли нужно использовать оба класса jdbc одновременно? (Фактически, с учётом предка, их три ).
Обратите внимание, что класс `NamedParameterJdbcTemplate` обертывает классический шаблон `JdbcTemplate`. Если вам нужен доступ к экземпляру обернутого `JdbcTemplate`, чтобы получить доступ к функциональности, которая присутствует только в классе `JdbcTemplate`, можно использовать метод `getJdbcOperations()` для получения доступа к обернутому `JdbcTemplate` через интерфейс `JdbcOperations`.

https://www.concretepage.com/spring-boot/jdbcclient-jdbctemplate-namedparameterjdbctemplate-spring-boot

https://javarush.com/quests/lectures/questspring.level03.lecture09

1Boolean1 replied:
Спасибо, исправлено, изучили всё подробнее

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/ReviewRepository.java:
```diff
@@ -0,0 +1,142 @@
+import java.util.Optional;
+
+/*
+Не стал оформлять ReviewDbStorage, так как в остальной программе такие классы лишь делигируют вызовы классу-репозиторию.
```
Честно говоря, я бы вообще избавился от слоя Storage. Часть методов можно перенести в соответствующие классы слоя репозитария, а часть вынести в слой сервисов. 

Не настаиваю, но, вообще говоря, желательно. Так, как это сделано сейчас, вносит некую запутанность в архитектуру проекта.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/review/ReviewService.java:
```diff
@@ -0,0 +1,195 @@
+                .orElseThrow(() -> new NotFoundException("Review with ID " + reviewId + " not found"));
+    }
+
+    @Transactional
```
Transactional

В принципе, для данного проекта можно обойтись без этой аннотации. )

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,56 @@
+@RestController
+@RequestMapping("/directors")
+@RequiredArgsConstructor
+@Validated
```
@Validated не имеет смысла, так как отсутствуют аннотации валидации(ограничения) для параметров методов)

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public DirectorDto addDirector(@RequestBody @Valid Director director) {
```
`    public DirectorDto addDirector(@RequestBody @Valid Director director) {`

Director director

Так, раз уж используется техника работы с ДТО классами, то давайте разбираться. 🙂
Здесь не должно быть объектов БД. Он же пришёл снаружи, значит должен быть dto.
Передаётся в слой сервиса как dto, там  происходит обработка, а перед тем как передать в слой репозитария, делается маппинг в объект базы данных.
Впрочем, маппинг можно делать и на входе в сервис. На Ваше усмотрение.
Ну и так далее, по остальным методам и остальным классам. Проверьте пожалуйста.
Общий принцип такой-контроллеры - это dto, сервисы это бизнес логика и маппинг, а репозиторий это обычные классы (сущности) для работы в базе данных.

1Boolean1 replied:
исправлено

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -56,15 +56,45 @@ public void addLikeToFilm(@PathVariable long filmId, @PathVariable long userId)
         log.info("Received DELETE /films/{}/like/{} request", filmId, userId);
         filmService.deleteLikeToFilm(filmId, userId);
     }
 
```
```
Validated
public class FilmController {
```
Аннотация над классом будет работать в том случае, если  к примитивным входным параметрам методов добавлены дополнительные ограничительные аннотации, например: 

`    public FilmDto getFilmById(@PathVariable  @Positive long filmId) {`

1Boolean1 replied:
исправлено

avfyodorov reviewed (commented):
Добрый день, Илья!

Вы отлично поработали над проектом. Функционал задач реализован в соответствии с техническим заданием,  в общем и целом разобрались с взаимодействием слоев Controller - service - repository (хотя тут ещё есть что поправить :)). Несмотря на то, что это первый групповой проект, результат можно считать очень хорошим. 

От меня только несколько уточнений, часть из них на ваше усмотрение.

Старался не  писать повторяющихся комментариев, будете править, проверьте, пожалуйста, аналогичные места и в других частях проекта.

Plastinin-Igor committed:
Исправление замечаний (#24)

* Добавить в файл README.md список выполненных задач

* Удалить класс filmorate/dto/UpdateFilmRequest.java

Класс FeedRepository унаследовать от BaseRepository
В классе GenreRepository убрать поля jdbc и genreMapper, в методе findGenresByFilmId использовать базовый метод findMany
В классе ReviewService убрать аннотацию Transactional
В классе DirectorController убрать аннотацию validated
В классе DirectorController входные параметры исправить с Director на DirectorDto
В классе DirectorService входные параметры исправить с Director на DirectorDto
В классе убрать аннотацию FilmController Validated

* Delete exceptions, refactor jdbc in FilmRepository

* checkstyle

* Использовать технику работы с DTO в контроллерах

* refactor

---------

Co-authored-by: Remenchik Ilya <ilyaremenchik228@gmail.com>

avfyodorov reviewed (commented):
Добрый день! 

Замечаний нет. 
Групповой проект принят. 
 Поздравляю с удачным опытом групповой разработки, успехов!
 Уверен, что данный опыт пригодится вам в дальнейшем.
